### PR TITLE
Changes required by Asukasbudjetti

### PR DESCRIPTION
**Pull request by riksu-raksu**
Asukasbudjetti requires

- Municipality and postal code from suomi.fi, to see which region the user is eligible to vote in
- A new login method (OPAS AD) that students in Turku can use.

OPAS AD can be used by some non-students as well (such as teachers), but only students are eligible to vote in Asukasbudjetti. This is why when OPAS AD is used to log in, the client application gets information about the role of the user ('school_role' attribute).

### CODE REVIEW

File: auth_backends/adfs/opas.py [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+
+
+class OpasADFS(SAMLAuth):
+    name = 'opas_adfs'
+    metadata_url = 'https://sts.edu.turku.fi/federationmetadata/2007-06/federationmetadata.xml'
+    EXTRA_DATA = ['school_role']
+
+    def generate_saml_config(self, idp=None):
+        ret = super().generate_saml_config(idp)
+        ret['security']['wantAssertionsSigned'] = True,
```
Comment (ezkat) [inline auth_backends/adfs/opas.py:31]: Do we want this to be a tuple or is the trailing comma a typo?

File: auth_backends/adfs/opas.py [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+    def get_user_details(self, response):
+        # get the default user details from the super class
+        # and add the 'role' among them
+
+        user_details = super().get_user_details(response)
+        idp = self.get_idp()
+
+        role = idp.get_attr(response['attributes'], 'attr_role', 'http://schemas.microsoft.com/ws/2008/06/identity/claims/role'),
+        if isinstance(role, tuple):
+            role = role[0]
```
Comment (ezkat) [inline auth_backends/adfs/opas.py:124]: Trailing comma will always make "role" a tuple, is if-statement required (or is the trailing comma necessary)?

Comment (riksu-raksu) [reply auth_backends/adfs/opas.py:124]: This comma is a typo, I'll remove it and the if-statement

File: auth_backends/adfs/opas.py (hunk 1 above)
Comment (riksu-raksu) [reply auth_backends/adfs/opas.py:31]: This comma I took straight from an example, after looking what's in `super().generate_saml_config()` I think it does not need to be there. I tried removing it and generating the config still works, so looks like the example I used has a typo in there.

**Review by ezkat — APPROVED** [review]

**Review by aceViilee — APPROVED** [review]

**Pull request merged**